dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -373,6 +375,77 @@ bool Segment::write(DMContext & dm_context, const RowKeyRange & delete_range)
     return delta->appendDeleteRange(dm_context, delete_range);
 }
 
+bool Segment::isDefinitelyEmpty(DMContext & dm_context, const SegmentSnapshotPtr & segment_snap) const
+{
+    RUNTIME_CHECK(segment_snap->isForUpdate());
+
+    // Fast path: all packs has been filtered away
+    if (segment_snap->getRows() == 0)
+    {
+        ProfileEvents::increment(ProfileEvents::DMSegmentIsEmptyFastPath);
+        return true;
+    }
+
+    ProfileEvents::increment(ProfileEvents::DMSegmentIsEmptySlowPath);
+
+    // Build read stream
+    auto columns_to_read = std::make_shared<ColumnDefines>();
+    columns_to_read->push_back(getExtraHandleColumnDefine(is_common_handle));
+    auto read_ranges = RowKeyRanges{rowkey_range};
+
+    SkippableBlockInputStreams streams;
+    for (const auto & file : segment_snap->stable->getDMFiles())
+    {
+        DMFileBlockInputStreamBuilder builder(dm_context.db_context);
+        auto stream = builder
+                          .setRowsThreshold(std::numeric_limits<UInt64>::max()) // TODO: May be we could have some better settings
```

Comment:
I think we don't need this line as we only try to read one row from the stream

Reply:
We may also read multiple rows in some rare cases, for example, the whole block may be filtered out by SegmentRange.